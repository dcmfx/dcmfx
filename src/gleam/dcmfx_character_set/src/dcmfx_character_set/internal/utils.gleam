import gleam/string

/// Returns the codepoint for the replacement character '�' that is emitted when
/// invalid string data is encountered.
///
pub fn replacement_character() -> UtfCodepoint {
  let assert Ok(codepoint) = string.utf_codepoint(0xFFFD)
  codepoint
}

/// Converts an integer codepoint value to a `UtfCodepoint`. The replacement
/// character '�' is returned if the integer is not a valid codepoint.
///
pub fn int_to_codepoint(i: Int) -> UtfCodepoint {
  case string.utf_codepoint(i) {
    Ok(codepoint) -> codepoint
    Error(Nil) -> replacement_character()
  }
}
