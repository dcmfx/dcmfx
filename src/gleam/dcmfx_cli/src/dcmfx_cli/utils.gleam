import gleam/result
import simplifile

/// Returns whether the specified file paths point to the same file on the same
/// device or volume once symlinks are resolved.
///
pub fn is_same_file(
  file0: String,
  file1: String,
) -> Result(Bool, simplifile.FileError) {
  use file0_info <- result.try(simplifile.file_info(file0))
  use file1_info <- result.try(simplifile.file_info(file1))

  Ok(file0_info == file1_info)
}
