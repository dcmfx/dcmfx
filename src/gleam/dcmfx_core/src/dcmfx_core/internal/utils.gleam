import gleam/bit_array
import gleam/float
import gleam/int
import gleam/list
import gleam/result
import gleam/string

/// Returns the length of a string in graphemes (on Erlang), or in UTF-16 code
/// units (on JavaScript).
///
/// This function is much faster than `string.length()` on JavaScript because it
/// uses `String.length()`, but be aware it returns a different result when the
/// string contains characters outside the Basic Multilingual Plane, or
/// graphemes made up of multiple codepoints.
///
/// This function should only be used where the above discrepancy will not occur
/// or does not matter and performance is more important.
///
@external(javascript, "../../dcmfx_core_ffi.mjs", "utils__string_fast_length")
pub fn string_fast_length(s: String) -> Int {
  string.length(s)
}

/// Pads a string to the desired length by prepending a pad string.
///
/// This implementation is faster than `string.pad_left()` on the JavaScript
/// target, but is not grapheme aware.
///
@external(javascript, "../../dcmfx_core_ffi.mjs", "utils__pad_start")
pub fn pad_start(s: String, desired_length: Int, pad_string: String) -> String {
  string.pad_left(s, desired_length, pad_string)
}

/// Helper function that parses a string to a float, handling the case where the
/// input string is actually an integer with no decimal point, or there are no
/// digits following the decimal point.
///
pub fn smart_parse_float(input: String) -> Result(Float, Nil) {
  let input = trim_right_codepoints(input, [0x2E])

  input
  |> float.parse
  |> result.lazy_or(fn() { float.parse(input <> ".0") })
}

/// Removes all occurrences of the specified characters from the end of a
/// string.
///
pub fn trim_right(s: String, chars: String) -> String {
  let codepoints =
    chars
    |> string.to_utf_codepoints
    |> list.map(string.utf_codepoint_to_int)

  trim_right_codepoints(s, codepoints)
}

/// Removes all whitespace from the end of the passed string. Whitespace is
/// defined as the following Unicode codepoints: U+0000, U+0009, U+000A, U+000D,
/// U+0020.
///
pub fn trim_right_whitespace(s: String) -> String {
  trim_right_codepoints(s, [0x00, 0x09, 0x0A, 0x0D, 0x20])
}

/// Removes all occurrences of the specified codepoints from the end of a
/// string. This function can only remove ASCII codepoints, i.e. those with
/// values <= `0x7F`.
///
pub fn trim_right_codepoints(s: String, codepoints: List(Int)) -> String {
  let s = bit_array.from_string(s)
  let len = bit_array.byte_size(s)

  do_trim_right_codepoints(s, len, codepoints)
}

fn do_trim_right_codepoints(
  s: BitArray,
  length: Int,
  codepoints: List(Int),
) -> String {
  case length {
    0 -> ""
    _ -> {
      // Get the last byte in the string
      let assert Ok(<<x>>) = bit_array.slice(s, length - 1, 1)

      case list.contains(codepoints, x) {
        True -> do_trim_right_codepoints(s, length - 1, codepoints)
        False -> {
          let assert Ok(s) = bit_array.slice(s, 0, length)
          let assert Ok(s) = bit_array.to_string(s)

          s
        }
      }
    }
  }
}

/// Index lookup into a list. This is used when looking up a specific sequence
/// item. This function was removed from the standard library in v0.38 because
/// of its O(N) performance.
///
pub fn list_at(in list: List(a), get index: Int) -> Result(a, Nil) {
  case index >= 0 {
    True ->
      case list_drop(list, index) {
        [] -> Error(Nil)
        [x, ..] -> Ok(x)
      }
    False -> Error(Nil)
  }
}

fn list_drop(from list: List(a), up_to n: Int) -> List(a) {
  case n <= 0 {
    True -> list
    False ->
      case list {
        [] -> []
        [_, ..xs] -> list_drop(xs, n - 1)
      }
  }
}

/// Inspects a bit array in hexadecimal, e.g. `[1A 2B 3C 4D]`.
///
pub fn inspect_bit_array(bits: BitArray) -> String {
  do_inspect_bit_array(bits, "[") <> "]"
}

fn do_inspect_bit_array(input: BitArray, accumulator: String) -> String {
  case input {
    <<x, rest:bytes>> -> {
      let suffix = case rest {
        <<>> -> ""
        _ -> " "
      }

      let accumulator =
        accumulator
        <> { x |> int.to_base16 |> string.pad_left(2, "0") }
        <> suffix

      do_inspect_bit_array(rest, accumulator)
    }

    _ -> accumulator
  }
}
