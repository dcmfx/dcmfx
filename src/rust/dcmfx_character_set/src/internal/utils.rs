/// The `char` for the replacement character '�' that is emitted when invalid
/// string data is encountered.
///
pub(crate) const REPLACEMENT_CHARACTER: char = '�';

/// Converts an integer codepoint value to a `char`. The replacement character
/// '�' is returned if the integer is not a valid codepoint.
///
pub(crate) fn codepoint_to_char(codepoint: u32) -> char {
  char::from_u32(codepoint).unwrap_or(REPLACEMENT_CHARACTER)
}
