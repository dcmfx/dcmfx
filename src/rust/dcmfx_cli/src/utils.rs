use std::{
  fs::File,
  io::Write,
  path::{Path, PathBuf},
  sync::{Arc, LazyLock, Mutex},
};

use dcmfx::p10::P10Error;

/// Creates a Rayon thread pool with the specified number of threads.
///
pub fn create_thread_pool(threads: usize) -> rayon::ThreadPool {
  rayon::ThreadPoolBuilder::new()
    .num_threads(threads)
    .build()
    .unwrap()
}

/// Shared stdout write stream that ensures only one thread writes at a time and
/// there's no unwanted interleaving of stdout output from multiple threads.
///
static GLOBAL_STDOUT: LazyLock<Arc<Mutex<Box<dyn Write + Send>>>> =
  LazyLock::new(|| Arc::new(Mutex::new(Box::new(std::io::stdout()))));

/// Opens an output stream for the given path, first checking whether it exists
/// and prompting the user about overwriting it if necessary. This prompt isn't
/// presented to the user if `overwrite` is true.
///
/// The path "-" is interpreted as writing to stdout.
///
pub fn open_output_stream(
  path: &PathBuf,
  display_path: Option<&PathBuf>,
  overwrite: bool,
) -> Result<Arc<Mutex<Box<dyn Write + Send>>>, P10Error> {
  if path.to_string_lossy() == "-" {
    Ok(GLOBAL_STDOUT.clone())
  } else {
    if let Some(display_path) = display_path {
      println!("Writing \"{}\" â€¦", display_path.display());
    }

    if !overwrite {
      error_if_exists(path);
    }

    match File::create(path) {
      Ok(file) => Ok(Arc::new(Mutex::new(Box::new(file)))),

      Err(e) => Err(P10Error::FileError {
        when: "Opening file".to_string(),
        details: e.to_string(),
      }),
    }
  }
}

/// Prints an error and exits the process if the specified file exists.
///
pub fn error_if_exists(path: &Path) {
  if !path.exists() {
    return;
  }

  eprintln!(
    "Error: Output file \"{}\" already exists. Specify --overwrite to
     automatically overwrite existing files",
    path.display()
  );
  std::process::exit(1);
}

/// Appends a suffix to a path.
///
pub fn path_append(mut path: PathBuf, suffix: &str) -> PathBuf {
  path.set_file_name(format!(
    "{}{}",
    path
      .file_name()
      .unwrap_or(std::ffi::OsStr::new(""))
      .to_string_lossy(),
    suffix
  ));

  path
}

/// Renames a temporary file to an output filename when [`Self::commit()`] is
/// called, otherwise deletes the temporary file on drop.
///
pub struct TempFileRenamer {
  temp_filename: PathBuf,
  output_filename: PathBuf,
  committed: bool,
}

impl TempFileRenamer {
  pub fn new(temp_filename: PathBuf, output_filename: PathBuf) -> Self {
    Self {
      temp_filename,
      output_filename,
      committed: false,
    }
  }

  pub fn commit(&mut self) -> Result<(), (String, String)> {
    self.committed = true;

    std::fs::rename(&self.temp_filename, &self.output_filename).map_err(|e| {
      (
        format!(
          "Renaming '{}' to '{}'",
          self.temp_filename.display(),
          self.output_filename.display()
        ),
        e.to_string(),
      )
    })
  }
}

impl Drop for TempFileRenamer {
  fn drop(&mut self) {
    if !self.committed {
      let _ = std::fs::remove_file(&self.temp_filename);
    }
  }
}
