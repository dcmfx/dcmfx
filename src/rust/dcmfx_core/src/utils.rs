#[cfg(feature = "std")]
use std::rc::Rc;

#[cfg(not(feature = "std"))]
use alloc::{format, rc::Rc, string::String, vec, vec::Vec};

/// A reference counted byte slice holds an `Rc<Vec<u8>>` to a shareable buffer,
/// along with a range that specifies the part of that buffer that this slice
/// refers to.
///
/// This type is used widely to avoid copying buffers wherever possible, and in
/// most cases can be used like a `&[u8]` would be.
///
#[derive(Clone)]
pub struct RcByteSlice {
  data: Rc<Vec<u8>>,
  range: core::ops::Range<usize>,
}

impl RcByteSlice {
  /// Creates a new referenced counted byte slice from a `Vec<u8>`.
  ///
  pub fn from_vec(data: Vec<u8>) -> Self {
    let range = 0..data.len();

    Self {
      data: Rc::new(data),
      range,
    }
  }

  /// Creates an empty reference counted byte slice.
  ///
  pub fn empty() -> Self {
    Self {
      data: Rc::new(vec![]),
      range: 0..0,
    }
  }

  /// Slices this referenced counted byte slice, returning a new reference
  /// counted byte slice that points to the same underlying data.
  ///
  /// This function does not copy any data.
  ///
  pub fn slice(&self, start: usize, end: usize) -> Self {
    assert!(start <= end, "Byte slice range out of bounds");
    assert!(end <= self.range.len(), "Byte slice range out of bounds");

    Self {
      data: self.data.clone(),
      range: (self.range.start + start)..(self.range.start + end),
    }
  }

  /// Returns a new reference counted byte slice with the specified number of
  /// bytes dropped from the front.
  ///
  pub fn drop(&self, n: usize) -> Self {
    self.slice(n, self.len())
  }

  /// Returns a new reference counted byte slice to the specified number of
  /// leading bytes.
  ///
  pub fn take(&self, n: usize) -> Self {
    self.slice(0, n)
  }

  /// Consumes this reference counted byte slice and turns it into a `Vec<u8>`.
  /// Avoids a copy when possible.
  ///
  /// This function copies data if there are multiple references to the
  /// underlying buffer, or its slice bounds do not cover the whole buffer.
  ///
  pub fn into_vec(self) -> Vec<u8> {
    if self.range == (0..self.data.len()) {
      match Rc::try_unwrap(self.data) {
        Ok(data) => data,
        Err(data_rc) => data_rc[self.range.clone()].to_vec(),
      }
    } else {
      self.as_slice().to_vec()
    }
  }

  fn as_slice(&self) -> &[u8] {
    &self.data[self.range.clone()]
  }
}

impl core::fmt::Debug for RcByteSlice {
  fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
    write!(f, "{:?}", self.as_slice())
  }
}

impl PartialEq for RcByteSlice {
  fn eq(&self, other: &Self) -> bool {
    self.as_slice() == other.as_slice()
  }
}

impl core::ops::Deref for RcByteSlice {
  type Target = [u8];

  fn deref(&self) -> &Self::Target {
    self.as_slice()
  }
}

impl From<Vec<u8>> for RcByteSlice {
  fn from(vec: Vec<u8>) -> Self {
    RcByteSlice::from_vec(vec)
  }
}

/// Inspects a byte slice in hexadecimal, e.g. `[1A 2B 3C 4D]`. If the number of
/// bytes in the slice exceeds `max_length` then not all bytes will be
/// shown and a trailing ellipsis will be appended, e.g. `[1A 2B 3C 4D …]`.
///
pub fn inspect_u8_slice(bytes: &[u8], max_length: usize) -> String {
  let byte_count = core::cmp::min(max_length, bytes.len());

  let s = bytes[0..byte_count]
    .iter()
    .map(|byte| format!("{:02X}", byte))
    .collect::<Vec<_>>()
    .join(" ");

  if byte_count == bytes.len() {
    format!("[{}]", s)
  } else {
    format!("[{} …]", s)
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[cfg(not(feature = "std"))]
  use alloc::string::ToString;

  #[test]
  fn inspect_u8_slice_test() {
    assert_eq!(
      inspect_u8_slice(&[0xD1, 0x96, 0x33], 100),
      "[D1 96 33]".to_string()
    );

    assert_eq!(
      inspect_u8_slice(&[0xD1, 0x96, 0x33, 0x44], 3),
      "[D1 96 33 …]".to_string()
    );
  }
}
