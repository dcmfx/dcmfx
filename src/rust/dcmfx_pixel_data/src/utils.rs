use core::ops::{Add, Div};

/// Unsigned integer division that rounds to the nearest integer instead of
/// truncating towards zero like standard integer division.
///
#[inline]
pub fn udiv_round<T>(numer: T, denom: T) -> T
where
  T: Copy + Add<Output = T> + Div<Output = T> + From<u8>,
{
  (numer + (denom / T::from(2))) / denom
}
