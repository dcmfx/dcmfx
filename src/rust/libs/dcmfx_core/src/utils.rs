/// Removes all whitespace from the end of the passed string. Whitespace is
/// defined as the following Unicode codepoints: U+0000, U+0009, U+000A, U+000D,
/// U+0020.
///
pub fn trim_right_whitespace(s: &str) -> &str {
  s.trim_end_matches([
    '\u{0000}', '\u{0009}', '\u{000A}', '\u{000D}', '\u{0020}',
  ])
}
