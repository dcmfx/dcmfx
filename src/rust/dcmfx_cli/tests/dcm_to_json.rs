mod utils;

use assert_cmd::Command;
use insta::assert_snapshot;
use utils::{generate_temp_filename, get_stdout};

#[test]
fn with_output_filename() {
  let dicom_file =
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd.dcm";
  let temp_path = generate_temp_filename();

  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();
  cmd
    .arg("dcm-to-json")
    .arg(dicom_file)
    .arg("--output-filename")
    .arg(&temp_path)
    .arg("--pretty")
    .assert()
    .success()
    .stdout(format!("Writing \"{}\" …\n", temp_path.display()));

  assert_snapshot!(
    "with_output_filename",
    std::fs::read_to_string(&temp_path).unwrap()
  );
}

#[test]
fn with_stdout_output() {
  let dicom_file =
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd.dcm";

  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();
  let assert = cmd
    .arg("dcm-to-json")
    .arg(dicom_file)
    .arg("--output-filename")
    .arg("-")
    .arg("--pretty")
    .assert()
    .success();

  assert_snapshot!("with_stdout_output", get_stdout(assert));
}

#[test]
fn with_output_directory() {
  let dicom_files = [
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd.dcm",
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd_jpeg.dcm",
  ];

  let output_directory = generate_temp_filename();
  std::fs::create_dir(&output_directory).unwrap();
  let output_files = [
    output_directory.join("SC_rgb_small_odd.dcm.json"),
    output_directory.join("SC_rgb_small_odd_jpeg.dcm.json"),
  ];

  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();
  cmd
    .arg("dcm-to-json")
    .args(dicom_files)
    .arg("--pretty")
    .arg("--output-directory")
    .arg(output_directory)
    .arg("--threads")
    .arg("1")
    .assert()
    .success()
    .stdout(format!(
      "Writing \"{}\" …\nWriting \"{}\" …\n",
      output_files[0].display(),
      output_files[1].display()
    ));

  assert_snapshot!(
    "with_output_directory_0",
    std::fs::read_to_string(&output_files[0]).unwrap()
  );
  assert_snapshot!(
    "with_output_directory_1",
    std::fs::read_to_string(&output_files[1]).unwrap()
  );
}

#[test]
fn with_multiple_inputs() {
  let dicom_files = [generate_temp_filename(), generate_temp_filename()];
  let output_files = [
    format!("{}{}", dicom_files[0].to_string_lossy(), ".json"),
    format!("{}{}", dicom_files[1].to_string_lossy(), ".json"),
  ];

  std::fs::copy(
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd.dcm",
    &dicom_files[0],
  )
  .unwrap();
  std::fs::copy(
    "../../../test/assets/pydicom/test_files/SC_rgb_small_odd_jpeg.dcm",
    &dicom_files[1],
  )
  .unwrap();

  let mut cmd = Command::cargo_bin("dcmfx_cli").unwrap();
  cmd
    .arg("dcm-to-json")
    .args(dicom_files)
    .arg("--pretty")
    .arg("--threads")
    .arg("1")
    .assert()
    .success()
    .stdout(format!(
      "Writing \"{}\" …\nWriting \"{}\" …\n",
      output_files[0], output_files[1]
    ));

  assert_snapshot!(
    "with_multiple_inputs_0",
    std::fs::read_to_string(&output_files[0]).unwrap()
  );
  assert_snapshot!(
    "with_multiple_inputs_1",
    std::fs::read_to_string(&output_files[1]).unwrap()
  );
}
