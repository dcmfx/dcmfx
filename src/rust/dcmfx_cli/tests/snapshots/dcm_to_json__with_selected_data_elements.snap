---
source: dcmfx_cli/tests/dcm_to_json.rs
expression: get_stdout(assert)
snapshot_kind: text
---
{"00080008":{"vr":"CS","Value":["DERIVED","SECONDARY","OTHER"]},"00080016":{"vr":"UI","Value":["1.2.840.10008.5.1.4.1.1.7"]}}
{"00080008":{"vr":"CS","Value":["DERIVED","SECONDARY","OTHER"]},"00080016":{"vr":"UI","Value":["1.2.840.10008.5.1.4.1.1.7"]}}
